core/src/kmx/kmx_actions.cpp

```diff
@@ -45,11 +45,12 @@ KMX_BOOL KMX_Actions::QueueAction(int ItemType, KMX_DWORD dwData)
     break;
 
   case QIT_CHAR:
-    if(Uni_IsSMP(dwData)) {
-      m_context->Add(Uni_UTF32ToSurrogate1(dwData));
-      m_context->Add(Uni_UTF32ToSurrogate2(dwData));
-    } else {
-      m_context->Add((KMX_WORD) dwData);
+    {
+      char16_single buf;
+      int len = Utf32CharToUtf16(dwData, buf);
+      for(int i=0; i<len; i++) {
+        m_context->Add(buf.ch[i]);
+      }
```

Comment:
Hmm... it's a shame that the replacement requires a loop. But not seeing a clean, simple way to avoid that. (I can see ugly C++ ways of avoiding the loop. But NO, we won't go there.)

Reply:
shouldn't the compiler optimize it?  'len' here is going to be 1 or 2 exactly.

Reply:
I would hope so. It's just a little verbose for code readers :grin:

Reply:
The other way to do this is to update Add() to take a string and length. Like I did with the .append() call.

Reply:
@mcdurdin  we could have a macro to collapse it like so:

```c
#define FOREACH_Utf32CharToUtf16(ch, var, action) { \
     char16_single buf;  \
     int len = Utf32CharToUtf16(ch, buf); \
     for(int i=0; i<len; i++) { \
        const char16_t &var = buf.ch[i]; /*alias*/ \
        (void)(action); \
     } \
  }
```

thus:

```c
  FOREACH_Utf32CharToUtf16(dwData, code_unit, m_context->add(code_unit));
```

… will turn into the same equivalent loop.

Reply:
yeah ... nah :grin: I'm also not a big fan of C macros burying logic..